### PR TITLE
separation of inheritances in yarpFixManager

### PR DESCRIPTION
**Pull request by aerydna**
YarpFixManager has been splitted in two parts, in order to ensure the destruction of the manager class to happen before the destruction of the yarp::os::Network object.
this is because the manager class depend to it.

### CODE REVIEW

File: src/libYARP_manager/src/manager.cpp [hunk 1]
```diff
@@ -937,7 +937,8 @@ void Manager::clearExecutables()
     for(itr=runnables.begin(); itr!=runnables.end(); itr++)
     {
         // broker will be deleted by Executable
-        delete (*itr);
```
Comment (traversaro) [inline src/libYARP_manager/src/manager.cpp]: Deleting a `nullptr` should be perfectly fine (https://stackoverflow.com/questions/4190703/is-it-safe-to-delete-a-null-pointer). Why was this necessary?

Comment (aerydna) [reply src/libYARP_manager/src/manager.cpp]: well.. reading the link you posted it seems that is still a good practice in case of delete operator overload. but since this is not the case i'll remove it.

File: src/rtf-plugins/fixture-managers/yarpmanager/YarpFixManager.cpp [hunk 2]
```diff
@@ -49,35 +50,31 @@ bool YarpFixManager::setup(int argc, char** argv) {
...
-        fixtureName = rf.find("fixture").asString();
-        std::string appfile = rf.findFileByName(std::string("fixtures/"+fixtureName).c_str());
+        manager = new YarpManagerPlugin(getDispatcher());
+        manager->fixtureName = rf.find("fixture").asString();
+
+
+        std::string appfile = rf.findFileByName(std::string("fixtures/"+manager->fixtureName).c_str());
         if (appfile.empty())
         {
             yInfo("yarpmanager: trying to load fixture file from absolute path");
```
Comment (traversaro) [inline src/rtf-plugins/fixture-managers/yarpmanager/YarpFixManager.cpp]: Not directly related to this, but if we support specifying the fixture with the absolute path I would not print something in case the user supplies an absoltue path.

**Comment by traversaro** [comment]
Similar to https://github.com/robotology/yarp/pull/1424#issuecomment-340534382 : which kind of bug is solved by this fix?

**Comment by aerydna** [comment]
this pr tries to solve a crash when using yarpserver fixture and yarpmanager fixture together.

File: src/libYARP_OS/src/ThreadImpl.cpp [hunk 3]
```diff
@@ -11,7 +11,7 @@
 #include <yarp/os/impl/PlatformSignal.h>
 #include <yarp/os/impl/PlatformThread.h>
 #include <yarp/os/impl/SemaphoreImpl.h>
-
```
[the remark this answers is not in the document]
Comment (drdanz) [reply src/libYARP_OS/src/ThreadImpl.cpp]: We already switched to use only c++ style headers, so please fix this

File: src/libYARP_OS/src/ThreadImpl.cpp [hunk 4]
```diff
@@ -359,6 +359,7 @@ bool ThreadImpl::isClosing()
 
 bool ThreadImpl::isRunning()
 {
+    assert(threadMutex);
```
Comment (drdanz) [inline src/libYARP_OS/src/ThreadImpl.cpp]: This assert does not make sense, it should be everywhere...
This mutex is protecting the access to a bool and to an int, instead these varables should be made atomic.

Comment (aerydna) [reply src/libYARP_OS/src/ThreadImpl.cpp]: i agree with you with making atomic the variables protected by this mutex.. however let me explain the reason for this assert:

the pointer is not checked before accessing it. there some situation where it is a nullptr and accessing it cause a crash. at first i've simply checked the pointer and returned false but @barbalberto rightly said that this kind of situation should not happen at all (a yarp thread is used without the "yarp::os::Network" singleton in memory) and the program should not continue... so for this specific case i thought an assert was better than a segfault.

**Comment by drdanz** [comment]
Merged, thanks.
